File: SOCKS/Benchmark/Socks5HandshakeParser.swift
import Foundation

struct Socks5HandshakeParser {
    static let maximumFeedBytes = 64 * 1024

    struct Target: Equatable {
        let address: String
        let port: UInt16
    }

    struct ConnectRequest: Equatable {
        let target: Target
        let firstPayload: Data
    }

    struct Output: Equatable {
        var replies: [Data] = []
        var connect: ConnectRequest?
        var shouldClose = false
    }

    private enum Phase {
        case greetingHeader
        case greetingMethods(Int)
        case requestHeader
        case requestIPv4Body
        case active
        case closed
    }

    private var phase = Phase.greetingHeader
    private var buffer = Data()

    var isActive: Bool {
        if case .active = phase { return true }
        return false
    }

    var isClosed: Bool {
        if case .closed = phase { return true }
        return false
    }

    var bufferedByteCount: Int {
        buffer.count
    }

    mutating func feed(_ data: Data) -> Output {
        guard data.count <= Self.maximumFeedBytes else {
            return closeForOversizedFeed()
        }
        guard !data.isEmpty, !isClosed else { return Output() }

        var output = Output()
        var offset = 0

        while offset < data.count, !output.shouldClose, output.connect == nil {
            switch phase {
            case .greetingHeader:
                guard appendNeeded(2, from: data, offset: &offset) else { break }
                let version = buffer[buffer.startIndex]
                let methodCount = Int(buffer[buffer.index(after: buffer.startIndex)])
                buffer.removeAll(keepingCapacity: true)

                guard version == 0x05, methodCount > 0 else {
                    output.replies.append(Data([0x05, 0xFF]))
                    close(&output)
                    continue
                }
                phase = .greetingMethods(methodCount)

            case .greetingMethods(let methodCount):
                guard appendNeeded(methodCount, from: data, offset: &offset) else { break }
                let supportsNoAuth = buffer.contains(0x00)
                buffer.removeAll(keepingCapacity: true)

                guard supportsNoAuth else {
                    output.replies.append(Data([0x05, 0xFF]))
                    close(&output)
                    continue
                }
                output.replies.append(Data([0x05, 0x00]))
                phase = .requestHeader

            case .requestHeader:
                guard appendNeeded(4, from: data, offset: &offset) else { break }
                let bytes = [UInt8](buffer)
                buffer.removeAll(keepingCapacity: true)

                let reply: UInt8?
                if bytes[0] != 0x05 {
                    reply = 0x01
                } else if bytes[2] != 0x00 {
                    reply = 0x01
                } else if bytes[1] != 0x01 {
                    reply = 0x07
                } else if bytes[3] != 0x01 {
                    reply = 0x08
                } else {
                    reply = nil
                }

                if let reply {
                    output.replies.append(Self.requestReply(reply))
                    close(&output)
                } else {
                    phase = .requestIPv4Body
                }

            case .requestIPv4Body:
                guard appendNeeded(6, from: data, offset: &offset) else { break }
                let bytes = [UInt8](buffer)
                buffer.removeAll(keepingCapacity: true)

                let address = "\(bytes[0]).\(bytes[1]).\(bytes[2]).\(bytes[3])"
                let port = UInt16(bytes[4]) << 8 | UInt16(bytes[5])
                let payloadStart = data.index(data.startIndex, offsetBy: offset)
                let payload = Data(data[payloadStart..<data.endIndex])
                offset = data.count
                phase = .active
                output.connect = ConnectRequest(
                    target: Target(address: address, port: port),
                    firstPayload: payload
                )

            case .active:
                offset = data.count

            case .closed:
                offset = data.count
            }
        }

        return output
    }

    mutating func finish() -> Output {
        guard !isClosed, !isActive else { return Output() }
        buffer.removeAll(keepingCapacity: false)
        phase = .closed
        return Output(shouldClose: true)
    }

    static func requestReply(_ reply: UInt8) -> Data {
        Data([0x05, reply, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00])
    }

    private mutating func appendNeeded(_ requiredCount: Int, from data: Data, offset: inout Int) -> Bool {
        let needed = requiredCount - buffer.count
        guard needed > 0 else { return true }

        let available = data.count - offset
        let count = min(needed, available)
        guard count > 0 else { return false }

        let start = data.index(data.startIndex, offsetBy: offset)
        let end = data.index(start, offsetBy: count)
        buffer.append(contentsOf: data[start..<end])
        offset += count
        return buffer.count == requiredCount
    }

    private mutating func close(_ output: inout Output) {
        buffer.removeAll(keepingCapacity: false)
        phase = .closed
        output.shouldClose = true
    }

    private mutating func closeForOversizedFeed() -> Output {
        var output = Output()
        switch phase {
        case .greetingHeader, .greetingMethods:
            output.replies.append(Data([0x05, 0xFF]))
        case .requestHeader, .requestIPv4Body:
            output.replies.append(Self.requestReply(0x01))
        case .active, .closed:
            break
        }
        close(&output)
        return output
    }
}

File: android/socks-core/src/main/kotlin/com/nanako/socksbypass/core/Socks5HandshakeParser.kt
package com.nanako.socksbypass.core

/**
 * Incremental SOCKS5 handshake parser (NO AUTH, CONNECT, UDP ASSOCIATE).
 * Parity with the Swift Socks5HandshakeParser.
 */
class Socks5HandshakeParser {
    data class Target(val address: String, val port: Int)

    enum class Command { Connect, UdpAssociate }

    data class Request(
        val command: Command,
        val target: Target,
        val firstPayload: ByteArray,
    ) {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is Request) return false
            return command == other.command &&
                target == other.target &&
                firstPayload.contentEquals(other.firstPayload)
        }

        override fun hashCode(): Int {
            var result = command.hashCode()
            result = 31 * result + target.hashCode()
            result = 31 * result + firstPayload.contentHashCode()
            return result
        }
    }

    data class Output(
        val replies: MutableList<ByteArray> = mutableListOf(),
        var request: Request? = null,
        var shouldClose: Boolean = false,
    )

    private enum class AddressKind { Ipv4, Ipv6, Domain }

    private sealed class Phase {
        data object GreetingHeader : Phase()
        data class GreetingMethods(val count: Int) : Phase()
        data object RequestHeader : Phase()
        data object RequestDomainLength : Phase()
        data class RequestAddress(val kind: AddressKind, val required: Int) : Phase()
        data object Active : Phase()
        data object Closed : Phase()
    }

    private var phase: Phase = Phase.GreetingHeader
    private var pendingCommand: Command = Command.Connect
    private val buffer = ArrayList<Byte>(64)

    val isActive: Boolean get() = phase is Phase.Active
    val isClosed: Boolean get() = phase is Phase.Closed
    val bufferedByteCount: Int get() = buffer.size

    fun feed(data: ByteArray): Output {
        if (data.size > MAXIMUM_FEED_BYTES) {
            return closeForOversizedFeed()
        }
        if (data.isEmpty() || isClosed) return Output()

        val output = Output()
        var offset = 0

        while (offset < data.size && !output.shouldClose && output.request == null) {
            when (val current = phase) {
                is Phase.GreetingHeader -> {
                    if (!appendNeeded(2, data, offsetRef = { offset }, setOffset = { offset = it })) break
                    val version = u8(buffer[0])
                    val methodCount = u8(buffer[1])
                    buffer.clear()
                    if (version != 0x05 || methodCount <= 0) {
                        output.replies.add(byteArrayOf(0x05, 0xFF.toByte()))
                        close(output)
                        continue
                    }
                    phase = Phase.GreetingMethods(methodCount)
                }

                is Phase.GreetingMethods -> {
                    if (!appendNeeded(current.count, data, offsetRef = { offset }, setOffset = { offset = it })) break
                    val supportsNoAuth = buffer.any { u8(it) == 0x00 }
                    buffer.clear()
                    if (!supportsNoAuth) {
                        output.replies.add(byteArrayOf(0x05, 0xFF.toByte()))
                        close(output)
                        continue
                    }
                    output.replies.add(byteArrayOf(0x05, 0x00))
                    phase = Phase.RequestHeader
                }

                is Phase.RequestHeader -> {
                    if (!appendNeeded(4, data, offsetRef = { offset }, setOffset = { offset = it })) break
                    val bytes = buffer.map { u8(it) }
                    buffer.clear()
                    val reply: Int? = when {
                        bytes[0] != 0x05 -> 0x01
                        bytes[2] != 0x00 -> 0x01
                        bytes[1] != 0x01 && bytes[1] != 0x03 -> 0x07
                        else -> null
                    }
                    val command = if (bytes[1] == 0x03) Command.UdpAssociate else Command.Connect
                    if (reply != null) {
                        output.replies.add(requestReply(reply))
                        close(output)
                    } else {
                        pendingCommand = command
                        when (bytes[3]) {
                            0x01 -> phase = Phase.RequestAddress(AddressKind.Ipv4, 6)
                            0x04 -> phase = Phase.RequestAddress(AddressKind.Ipv6, 18)
                            0x03 -> phase = Phase.RequestDomainLength
                            else -> {
                                output.replies.add(requestReply(0x08))
                                close(output)
                            }
                        }
                    }
                }

                is Phase.RequestDomainLength -> {
                    if (!appendNeeded(1, data, offsetRef = { offset }, setOffset = { offset = it })) break
                    val length = u8(buffer[0])
                    buffer.clear()
                    if (length <= 0) {
                        output.replies.add(requestReply(0x01))
                        close(output)
                        continue
                    }
                    phase = Phase.RequestAddress(AddressKind.Domain, length + 2)
                }

                is Phase.RequestAddress -> {
                    if (!appendNeeded(current.required, data, offsetRef = { offset }, setOffset = { offset = it })) break
                    val bytes = buffer.map { u8(it) }
                    buffer.clear()
                    val addressPart = bytes.subList(0, current.required - 2)
                    val address = decodeAddress(current.kind, addressPart)
                    val port = (bytes[current.required - 2] shl 8) or bytes[current.required - 1]
                    val payload = if (offset < data.size) data.copyOfRange(offset, data.size) else ByteArray(0)
                    offset = data.size
                    phase = Phase.Active
                    output.request = Request(
                        command = pendingCommand,
                        target = Target(address = address, port = port),
                        firstPayload = payload,
                    )
                }

                is Phase.Active, is Phase.Closed -> {
                    offset = data.size
                }
            }
        }
        return output
    }

    fun finish(): Output {
        if (isClosed || isActive) return Output()
        buffer.clear()
        phase = Phase.Closed
        return Output(shouldClose = true)
    }

    private fun appendNeeded(
        requiredCount: Int,
        data: ByteArray,
        offsetRef: () -> Int,
        setOffset: (Int) -> Unit,
    ): Boolean {
        val needed = requiredCount - buffer.size
        if (needed <= 0) return true
        val offset = offsetRef()
        val available = data.size - offset
        val count = minOf(needed, available)
        if (count <= 0) return false
        for (i in 0 until count) {
            buffer.add(data[offset + i])
        }
        setOffset(offset + count)
        return buffer.size == requiredCount
    }

    private fun close(output: Output) {
        buffer.clear()
        phase = Phase.Closed
        output.shouldClose = true
    }

    private fun closeForOversizedFeed(): Output {
        val output = Output()
        when (phase) {
            is Phase.GreetingHeader, is Phase.GreetingMethods ->
                output.replies.add(byteArrayOf(0x05, 0xFF.toByte()))
            is Phase.RequestHeader, is Phase.RequestDomainLength, is Phase.RequestAddress ->
                output.replies.add(requestReply(0x01))
            else -> Unit
        }
        close(output)
        return output
    }

    companion object {
        const val MAXIMUM_FEED_BYTES = 64 * 1024

        fun requestReply(reply: Int): ByteArray =
            byteArrayOf(0x05, reply.toByte(), 0x00, 0x01, 0, 0, 0, 0, 0, 0)

        data class BoundEndpoint(
            val atyp: Int,
            val address: ByteArray,
            val port: Int,
        )

        fun requestReply(reply: Int, bound: BoundEndpoint?): ByteArray {
            if (bound == null) return requestReply(reply)
            val expected = if (bound.atyp == 0x01) 4 else 16
            if (bound.address.size != expected) return requestReply(reply)
            val out = ByteArray(4 + bound.address.size + 2)
            out[0] = 0x05
            out[1] = reply.toByte()
            out[2] = 0x00
            out[3] = bound.atyp.toByte()
            System.arraycopy(bound.address, 0, out, 4, bound.address.size)
            val portIndex = 4 + bound.address.size
            out[portIndex] = ((bound.port shr 8) and 0xFF).toByte()
            out[portIndex + 1] = (bound.port and 0xFF).toByte()
            return out
        }

        private fun u8(b: Byte): Int = b.toInt() and 0xFF

        private fun decodeAddress(kind: AddressKind, bytes: List<Int>): String = when (kind) {
            AddressKind.Ipv4 -> bytes.joinToString(".") { it.toString() }
            AddressKind.Ipv6 -> bytes.chunked(2).joinToString(":") { pair ->
                String.format("%02x%02x", pair[0], pair[1])
            }
            AddressKind.Domain -> String(bytes.map { it.toByte() }.toByteArray(), Charsets.UTF_8)
        }
    }

    private fun u8(b: Byte): Int = b.toInt() and 0xFF

    private fun decodeAddress(kind: AddressKind, bytes: List<Int>): String =
        Companion.decodeAddress(kind, bytes)
}
